README.md
#SS_Servorator

An Arduino library that provides speed control for servos. Currently only the velocity of movement is controlled. Acceleration control will be added soon.

Servorator is written to be completely modular so that it can be 
used with any servo interface. This also means that Servorator requires
a little more work to set up and use.

While Servorator operates purely within the logical angle range of 0 to 180, you can use the [**SS_ServoTrim**](https://github.com/solderspot/SS_ServoTrim) library to simplify adding support for fine tuning the actual physical pulse timings for each servo. That library also provides the ability to dynamically remap and associate data with each servo. See [SimpleServoTrim.ino](examples/SimpleServoTrim/SimpleServoTrim.ino) for an example.

##How To Use

In your sketch you need to include the library so:


	#import <SS_Servorator.h>

Create an instance of SS_Servorator. Pass in the number of servos to control:

	#define NUM_SERVOS  6

	SS_Servorator sr(NUM_SERVOS);

You can create multiple instances if you needed to manage groups of servos.

###Set Up

Each instance requires a servo handling routine of type **SS_ServoHandler**:

	// a servo handler for Servorator
	void update_servo( SS_Index index, SS_Angle angle, void *data)
	{
  		// SS_Angle is in 1000ths of a degree
  		servo[index].write( angle/1000);
	}

The routine is responsible for making the appropriate calls to whatever servo driver interface you are using. In the code above we are calling the Arduino's Servo library write() function. 

**SS_Index** is a servo index, a value between 0 and *NUM_SERVOS-1*. **SS_Angle** is in 1000ths of degrees and has the range **0 to 180000** (180 degrees). 

You can get the number of servos using **numServos()**:

	int num = sr.numServos();

The servo handler routine needs to be registered with the Servorator instance using **setServoHandler()**:

	void setup()
	{
  		:
  		// register servo handler
  		// void setServoHandler( SS_ServoHandler *handler, void *user_data)
  		sr.setServoHandler( update_servo, NULL);
  		:	
	}

The second parameter is a pointer to void. Whatever pointer you pass will in turn be passed through to the update routine in the 'data' param shown above. You can use this param to associate user data with the handler.

For the handler to get called you need to regularly call Servorator's **service()** function:

	// main loop
	void loop()
	{
  		:
  		sr.service();
  		:
	}
	
If service() is not called frequently servo movement will become erratic.

###Control

To move a servo to a new postion use **setServoTargetAngle()**:

	// void setServoTargetAngle( SS_Index servo, SS_Angle new_angle )
    sr.setServoTargetAngle( 0, SS_DEGREES(175));

The **SS_DEGREES()** macro helps make coding a little more readable. You can also type SS_DEGREES(175.5) but keep in mind that Servorator is only accurate to a 1000th of a degree so SS_DEGREES(175.572987) is identical to SS_DEGREES(175.572).

The max speed at which the servos will turn is set using **setServoMaxVelocity()**:

	// void setServoMaxVelocity( SS_Index servo, SS_Velocity vel )
    sr.setServoMaxVelocity( 0, SS_DEGREES(60));

**SS_Velocity** is in 1000ths of degrees per second.

###Query Status

You can read the current and target servo angles using **getServoAngle** and **getServoTargetAngle()**:

	// SS_Angle getServoAngle( SS_Index servo )
    SS_Angle current = sr.getServoAngle( 0 );
    
	// SS_Angle getServoTargetAngle( SS_Index servo )
    SS_Angle target = sr.getServoTargetAngle( 0 );

'target' is the angle set in the last call to setServoTargetAngle(). 'current' is the angle the servo is currently at.

##Example

The library includes the **[SimpleServo.ino](examples/SimpleServo/SimpleServo.ino)** example:

    // include needed librarys
    #include <Servo.h>
    #include <SS_Servorator.h>


    #define NUM_SERVOS 6

    Servo servo[NUM_SERVOS];
    SS_Servorator sr(NUM_SERVOS);

    // the servo handler for Servorator
    void update_servo( SS_Index index, SS_Angle angle, void *data)
    {
      // SS_Angle is in 1000th of a degree
      servo[index].write( angle/1000);
    }

    void setup() 
    {

      // assign PWM pins to servos
      servo[0].attach(3);
      servo[1].attach(5);
      servo[2].attach(6);
      servo[3].attach(9);
      servo[4].attach(10);
      servo[5].attach(11);
      
      // register servo handler
      sr.setServoHandler( update_servo, NULL);

      // set all servos at 45 degrees
      // servo 0 is fastest
      SS_Velocity vel = SS_FAST_RATE;
      for ( int i=0; i<NUM_SERVOS;i++)
      {
        sr.setServoTargetAngle(i, SS_DEGREES(45)); 
        sr.setServoMaxVelocity( i, vel );
        vel = vel / 2 ;
      }

    }

    // main loop
    void loop()
    {
      // make servos ping-pong between 45 and 135 degrees
      for ( int i=0; i<NUM_SERVOS;i++)
      {
        SS_Angle angle = sr.getServoAngle(i); 

        if ( angle <= SS_DEGREES(45))
        { 
          sr.setServoTargetAngle(i, SS_DEGREES(135));
        } 
        else if ( angle >= SS_DEGREES(135))
        { 
          sr.setServoTargetAngle(i, SS_DEGREES(45));
        } 
      }
      // sr.service() needs to be called regularly so that
      // the servos are updated via update_servo()
      sr.service();
    }


##Other Examples

The library also includes these examples.

 * **[SimpleServoTrim.ino](examples/SimpleServoTrim/SimpleServoTrim.ino)** - A version of the SimpleServo.ino example that uses the **SS_ServoTrim** library to adjust servo pulse ranges and perform servo mappings.
 
 * **[AdafruitTrim.ino](examples/AdafruitTrim/AdafruitTrim.ino)** - A version of SimpleServoTrim.ino that uses Adafruit's servo shield to drive the servos.



changelog.md
#Change Log for SS_Servorator

**V1.2.0** - 1/2/14

 * Added SS_ServoTrim support
 * Added ServoTrim examples


SS_Servorator.cpp
//
//  
//  SS_Servorator.cpp
//
// Copyright (c) 2013-2014, Solder Spot
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
// 
// * Redistributions of source code must retain the above copyright notice, this
//   list of conditions and the following disclaimer.
// 
// * Redistributions in binary form must reproduce the above copyright notice, this
//   list of conditions and the following disclaimer in the documentation and/or
//   other materials provided with the distribution.
// 
// * Neither the name of the copyright holders nor the names of its
//   contributors may be used to endorse or promote products derived from
//   this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#if ARDUINO >= 100
 #include "Arduino.h"
#else
 #include "WProgram.h"
#endif

#include <stdlib.h> // for malloc and free

#include "SS_Servorator.h"


struct SS_Servo
{
     SS_Index               index;

     SS_Velocity            max_velocity; // max veocity allowed

     SS_Velocity            velocity;       // current velocity vector (sign means direction)
     SS_Acceleration        acceleration;   // acceleration

     SS_Angle               current_angle;      // angle servo is currently positioned at
     SS_Angle               target_angle;       // angle we want the servo to reach

     SS_Time                last_update;        // time of last update
     SS_Time                a_time;             // accelration time
     SS_Time                s_time;             // sustain velocity time
     SS_Time                d_time;             // deacceleration time
     SS_Time                b_time;             // time to break (to change direction)

};

//----------------------------------------
//
//----------------------------------------

SS_Servorator::SS_Servorator(int count)
{
    // create servo array
    _servo = (SS_Servo*)malloc(sizeof(SS_Servo)*count);
    _num_servos = count;

    // init servo data
    SS_Servo *s = _servo;
    for ( int i=0; i<_num_servos;i++,s++)
    {
        s->index = i;
        s->max_velocity = SS_NORMAL_RATE;
        s->acceleration = 0;
        s->current_angle = SS_NO_ANGLE;
        s->target_angle = SS_NO_ANGLE;
        s->last_update = 0;
        stopped(s);
    }
    _handler = NULL;
    _handler_data = NULL;
    _service_interval = 20;
    _last_service_time = 0;

}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::setServoHandler( SS_ServoHandler *handler, void *data)
{
    _handler = handler;
    _handler_data = data;
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::setServoTargetAngle( SS_Index index, SS_Angle new_angle)
{
    if ( index == SS_ALL)
    {
        for (int i=0;i<_num_servos;i++ )
        {
            setServoTargetAngle( i, new_angle);
        }

        return;
    }

    SS_Servo *servo = get_servo(index);
    if ( servo && servo->target_angle != new_angle )
    {
        servo->target_angle = new_angle;
        update_accel( servo );
    }
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::setServoMaxVelocity( SS_Index index, SS_Velocity new_max )
{
    if ( new_max < 0)
    {
        return;
    }

    if (index == SS_ALL) 
    {
        for (int i=0;i<_num_servos;i++ )
        {
            setServoMaxVelocity( i, new_max);
        }

        return;
    }

    SS_Servo *servo = get_servo(index);
    if ( servo && servo->max_velocity != new_max)
    {
        servo->max_velocity = new_max;
        update_accel( servo);
    }
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::setServoAcceleration( SS_Index index, SS_Acceleration new_acc )
{
    if ( index == SS_ALL)
    {
        for (int i=0;i<_num_servos;i++ )
        {
            setServoAcceleration( i, new_acc);
        }

        return;
    }

    SS_Servo *servo = get_servo(index); 
    if ( servo && servo->acceleration != new_acc)
    {
        servo->acceleration = new_acc;
        update_accel( servo);
    }
}

//----------------------------------------
//
//----------------------------------------

SS_Angle SS_Servorator::getServoAngle( SS_Index index)
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return servo->current_angle;
    }

    return SS_NO_ANGLE;
}

//----------------------------------------
//
//----------------------------------------

SS_Angle SS_Servorator::getServoTargetAngle( SS_Index index)
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return servo->target_angle;
    }

    return SS_NO_ANGLE;
}

//----------------------------------------
//
//----------------------------------------

SS_Velocity SS_Servorator::getServoMaxVelocity( SS_Index index )
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return servo->max_velocity;
    }

    return 0;
}

//----------------------------------------
//
//----------------------------------------

SS_Velocity SS_Servorator::getServoVelocity( SS_Index index )
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return servo->velocity;
    }

    return 0;
}

//----------------------------------------
//
//----------------------------------------

SS_Acceleration SS_Servorator::getServoAcceleration( SS_Index index )
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return servo->acceleration;
    }

    return 0;
}

//----------------------------------------
//
//----------------------------------------

SS_Time SS_Servorator::getServoTimeTillTarget( SS_Index index )
{
    SS_Servo *servo = get_servo(index);
    if ( servo )
    {
        return time_till_target( servo );
    }

    return 0;
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::update_servo( SS_Servo *servo, SS_Time now )
{

    if ( servo->current_angle == servo->target_angle)
    {
        // nothing to do
        return;
    }

    if ( servo->target_angle < 0)
    {
        // no angle set
        servo->current_angle = servo->target_angle;
        servo->last_update = now;
        stopped(servo);
        return;
    }

    if ( servo->current_angle < 0 )
    {
        // we don't have a current angle so go directly to the target position
        servo->current_angle = servo->target_angle;
        servo->last_update = now;
        stopped(servo);
        _handler( servo->index, servo->current_angle, _handler_data);
        return;
    }

    // we have an angle we need to move the servo to
    // however we need to move at the appropriate rate

    signed long delta = (signed long) (now - servo->last_update);
    
    SS_Angle increment = (servo->velocity*delta)/1000L;
    //Serial.println(increment);
    if ( increment != 0 )
    {
        servo->current_angle += increment;
        servo->last_update = now;

        // clip the change
        if ( increment > 0 )
        {
            servo->current_angle = servo->current_angle > servo->target_angle ? servo->target_angle : servo->current_angle; 
        }
        else
        {
            servo->current_angle = servo->current_angle < servo->target_angle ? servo->target_angle : servo->current_angle; 
        }

        if ( servo->current_angle == servo->target_angle)
        {
            stopped(servo);
        }
        _handler( servo->index, servo->current_angle, _handler_data);
    }

    if ( servo->s_time )
    {
        servo->s_time -= delta > servo->s_time ? servo->s_time : delta;
    }
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::update_accel( SS_Servo *servo  )
{
    //Serial.println("update_accel");
    SS_Angle diff = servo->target_angle - servo->current_angle;

    if ( diff == 0 || servo->max_velocity == 0)
    {
        // trivial case
        stopped( servo );
        return;
    }

    long dir = diff < 0 ? -1 : 1;
    SS_Angle mag = diff < 0 ? -diff : diff;

    if ( servo->acceleration == 0)
    {
        // trivial case
        //Serial.println("accel = 0");
        servo->velocity = servo->max_velocity*dir;
        servo->a_time = servo->d_time = servo->b_time = 0;
        servo->s_time = servo->velocity != 0 ? (mag*1000L)/(servo->velocity < 0 ? -servo->velocity : servo->velocity) : 0;
        //Serial.print("velocity = ");
        //Serial.println(servo->velocity);
        //Serial.print("time = ");
        //Serial.println(servo->s_time);
        return;
    }

    #if 0
    SS_

    if ( servo->velocity)
    {
        // we are already in motion so we need to see if we can 
    }
    SS_Angle a_move;
    SS_Angle b_move;

    servo->a_time = calc_deltas( servo->max_velocity - servo->velocity, servo->acceleration, &a_move );
    servo->d_time = calc_deltas( servo->max_velocity, servo->acceleration, &d_move );

    if (a_move + b_move <= mag )
    {
        // trival case
        servo->s_time = ((mag-a_move-d_move)*1000L)/servo->max_velocity;
        return;
    }

    // not enough space to perform full accel/deaccel

    if (servo->velocity == 0) 
    {
        // just split the remaing space between accel and deaccel
        servo->a_time = servo->b_time = calc_accel_time( 0, servo->acceleration, mag/2);
        servo->deacceleration = servo->acceleration;
        servo->s_time = 0;
        return;

    }

    if ( servo->velocity == servo->max_velocity )
    {
        // just deaccel
        servo->deacceleration = calc_accel( servo->velocity, 0, mag, &servo->d_time);

    }
    // calc the movement needed to perform normal deaccel from current velocity
    // this is the hard one....
    #endif

}

//----------------------------------------
//
//----------------------------------------

SS_Angle SS_Servorator::time_till_target( SS_Servo *servo  )
{
    return servo->a_time + servo->s_time + servo->d_time + servo->b_time;
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::stopped( SS_Servo *servo  )
{
    servo->velocity = 0;
    servo->a_time = servo->b_time = servo->s_time = servo->d_time = 0;
}


//----------------------------------------
//
//----------------------------------------

void SS_Servorator::service( void )
{
    SS_Servo *s = _servo;
    SS_Time now = millis();
    SS_Time delta = now - _last_service_time;

    if ( delta >= _service_interval )
    {
        for (int i=0; i<_num_servos; i++)
        {
            update_servo( s++, now );
        }

        _last_service_time = now;
    }
}

//----------------------------------------
//
//----------------------------------------

void SS_Servorator::setUpdateInterval( SS_Time interval)
{
    _service_interval = interval > 1 ? interval : 1;
}

//----------------------------------------
//
//----------------------------------------

SS_Servo *SS_Servorator::get_servo( SS_Index index)
{
    if ( index >= 0 && index < _num_servos && _servo )
    {
        return _servo+index;
    }

    return NULL;
}




SS_Servorator.h
//
//
// SS_Servorator.h
//
// Copyright (c) 2013-2014, Solder Spot
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
// 
// * Redistributions of source code must retain the above copyright notice, this
//   list of conditions and the following disclaimer.
// 
// * Redistributions in binary form must reproduce the above copyright notice, this
//   list of conditions and the following disclaimer in the documentation and/or
//   other materials provided with the distribution.
// 
// * Neither the name of the copyright holders nor the names of its
//   contributors may be used to endorse or promote products derived from
//   this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


#ifndef _SS_Servorator_H
#define _SS_Servorator_H

#define SS_ServoratorVersion        010200          // V1.1.0

struct SS_Servo;     // private struct

// in Servorator all angles are in 1000th of degrees.

typedef long SS_Velocity;       // in 1000th of degrees per second
typedef long SS_Acceleration;   // in 1000th of degrees per second per second
typedef long SS_Angle;          // in 1000th of degrees
typedef int SS_Index;           // value in range 0..numServos()-1
typedef unsigned long SS_Time;  // time in ms since startup

// servo handler is a user provide function that handles updating servo positions
// <servo> is index of the servo. <new_angle> is the new position of the servo in 1000th of degrees
// <data> is the point that was passed into SS_Servorator::setServoHandler()
typedef void SS_ServoHandler( SS_Index servo, SS_Angle new_angle, void *data);


// useful macros
#ifndef SS_DEGRESS
#define SS_DEGREES(D)        ((long)(((long)(D))*1000L))
#endif

#define SS_FAST_RATE    SS_DEGREES(180)             // 180 degrees per second
#define SS_NORMAL_RATE  SS_DEGREES(60)              // default rate - 60 dps
#define SS_SLOW_RATE    SS_DEGREES(15)              // 15 dps

#define SS_FAST_ACCEL   SS_DEGREES(180)             // 180 dpsps
#define SS_NORMAL_ACCEL SS_DEGREES(60)              // 18 dpsps
#define SS_SLOW_ACCEL   SS_DEGREES(15)              // 15 dpsps

#define SS_NO_ANGLE     -1      
#ifndef SS_ALL
#define SS_ALL          -1
#endif

class SS_Servorator 
{

 public:

  // create servorator with num_servo instacnes
  SS_Servorator(int num_servos);

  void setServoHandler( SS_ServoHandler *handler, void *data); // handler to call to set PWM signal for individual servos
  void setUpdateInterval( SS_Time interval);    // default is 20ms (50hz)

  // individual servo functions - servo index must be between 0..numServos()-1
  // setters
  void setServoTargetAngle( SS_Index servo, SS_Angle new_angle);        // set new target angle for servo
  void setServoMaxVelocity ( SS_Index servo, SS_Velocity vel );         // set new max velocity for servo
  void setServoAcceleration( SS_Index servo, SS_Acceleration acc );     // set new acceleration for servo

  // getters
  SS_Angle          getServoTargetAngle( SS_Index servo);               // get servo's current target angle
  SS_Velocity       getServoMaxVelocity( SS_Index servo );              // get servo's current max velocity
  SS_Velocity       getServoVelocity( SS_Index servo );                 // get servo's current velocity
  SS_Angle          getServoAngle( SS_Index servo);                     // get servo's current angle 
  SS_Acceleration   getServoAcceleration( SS_Index servo );             // get servo's current acceleration
  SS_Time           getServoTimeTillTarget( SS_Index servo);            // returns the estimated time till target angle is reached

  // other
  int numServos( void ) { return _num_servos;}
  void service( void );                                                 // service and update all servos as needed


 private:

  SS_Servo          *_servo;
  int               _num_servos;
  SS_ServoHandler   *_handler;
  void              *_handler_data;
  SS_Time           _last_service_time;
  SS_Time           _service_interval;

  SS_Servo          *get_servo(SS_Index index);
  void              update_servo( SS_Servo *servo, SS_Time now );
  void              update_accel( SS_Servo *servo );
  void              stopped( SS_Servo *servo );
  SS_Angle          time_till_target( SS_Servo *servo );

};

#endif

